src/router/routes/tool/api/topcards.ts

```diff
@@ -0,0 +1,74 @@
+import Joi from 'joi';
+
+import { makeFilter } from '../../../../client/filtering/FilterCards';
+import { SortDirections, SortDirectionsType } from '../../../../client/hooks/UseSortableData';
+import { OrderedSortsType, SortFunctions } from '../../../../client/utils/Sort';
+import { csrfProtection } from '../../../../routes/middleware';
+import { NextFunction, Request, Response } from '../../../../types/express';
+import { searchCards } from '../../../../util/tools';
+
+const validSortNames = Object.keys(SortFunctions);
+
+const TopCardsQuerySchema = Joi.object({
+  f: Joi.string().empty('').optional(),
+  s: Joi.string()
+    .valid(...validSortNames)
+    .required(),
+  d: Joi.string()
+    .valid(...SortDirections)
+    .required(),
+  p: Joi.number().integer().min(0).required(),
+}).unknown(true); // allow additional fields
+
+const validateQuery = (req: Request, res: Response, next: NextFunction) => {
```

Comment:
I'm not sure how best to test this. It wouldn't be a unit test as much as an integration test, since we want to test the rules against the request for real.